### ISSUE TITLE
How to implement fine-tuning phase?

### ISSUE THREAD

**Issue by xiexijun** [Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection#15]
The approaches in pervious papers like Meta R-CNN all have fine-tuning phase. Can you tell me how to implement this phase with your code? Looking forward to your reply!

**Comment by pengyinxw** [comment]
Hi, I also wonder how to implement fine-tuning phase, since you have published inference results using fine-tuned dana model. 
Furthermore,  it seems that you didn't use BA block when training base model as shown [here](https://github.com/Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection/blob/cad5632e08c8e300c95425b6eef6aa8671bc862a/lib/model/framework/dana.py#L328), semantic_enhance is always False.

**Comment by Tung-I** [comment]
1. To fine-tune the model, you should prepare a json file that has 10 or 30-shot images for each object class with corresponding box annotations and resume training on the fine-tuning dataset with a smaller learning rate. Using the same FewShotLoader would be fine. For your convenience, here are the json files I used for fine-tuning: 

https://drive.google.com/file/d/1eUZpc6KpSouZm8QL5s2EHXi4sXJXn89X/view?usp=sharing (10 shot)
https://drive.google.com/file/d/1zCj4Sbhd2FjHVxlmlk4Nfb9qqJcqMHRy/view?usp=sharing (30 shot)

2. Thank you for pointing out the issue of semantic_enhance=False. Currently, the only fine-tuned model weights I preserved  in hand is without BA block, but I am busy with a totally different project and have no time re-training a model to publish new weights. To obtain a completely well-trained DAnA mode, you could set semantic_enhance=True at utils.py and train the model on the json files either I provided or constructed by yourself.

**Comment by pengyinxw** [comment]
Thank you so much for your reply and provided json files. I am wondering how can I generate the fine-tuning dataset by myself. Could you please also provide the generation scripts and running commands for fine-tuning, if possible.

**Comment by pengyinxw** [comment]
Hi, when I try to fine-tune the model using your provided fine-tuning datasets(json files), it raises a value error: Sample larger than population or is negative. The problem comes from [here](https://github.com/Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection/blob/eacee935b6fe2a1fd572fa457eb39ae5d6dadb8c/lib/roi_data_layer/fs_loader.py#L112). After checking the [support_db](https://github.com/Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection/blob/eacee935b6fe2a1fd572fa457eb39ae5d6dadb8c/lib/roi_data_layer/fs_loader.py#L59), I found that some of the novel categories don't have enough shots(10/30shots) after the [box filter](https://github.com/Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection/blob/eacee935b6fe2a1fd572fa457eb39ae5d6dadb8c/lib/roi_data_layer/fs_loader.py#L75), while some have more than 10/30 shots. I am wondering how did you perform the fine-tuning step with these datasets and FewShotLoader?
One possible solution would replace the [random.sample method](https://github.com/Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection/blob/eacee935b6fe2a1fd572fa457eb39ae5d6dadb8c/lib/roi_data_layer/fs_loader.py#L112) with random.choices(), but in this way, it actually violates the N-way 10/30-shots setting.

For those categories which have more than 10/30 shots, if we random.sample 10/30 shots from them for each query image, then we will use more than 10/30 shots after iterating all the query images. it will also violate the N-way 10/30-shots setting.

**Comment by xiexijun** [comment]
Thank you very much for answering my question and providing the json files. I also met the same problem as @pengyinxw  when fine-tuning. Another question is how to set N and K of the N-way K-shot used in training phase in Table 1.

**Comment by Tung-I** [comment]
@pengyinxw  You ask a good question that bothered me a lot when doing this work. As I stated in Sec.4 B-3 (Experimental Settings-FSOD Evaluation) of the paper, different works adopted different evaluation protocols, which is rather annoying. You can regard the "shot" as either the number of box annotations or the number of query images. Most of the previous works seem to consider it as the number of query images or didn't clearly identify it. Therefore, to compare with the reported performance of previous works, I fine-tuned the model with 10/30 novel images per novel class. I assume the error you faced resulted from that you set shot=10/30 at fine-tuning (re-training). Please note that I still set shot=3 or 5 at fine-tuning since 10/30 indicates the amount of available novel data rather than the amount of support images your should fed to 
the model.

**Comment by pengyinxw** [comment]
@Tung-I Thanks  for your reply! As you assumed, I actually set shot =10/30 when fine-tuning and regard the "shot" as the number of box annotations instead of images.

As you mentioned above, you regard the "shot" as number of images in your experiments, but in your publication Sec.4 B-3, you write "we follow the widely-used protocol in [1] and compare ...", here the reference [1] actually regards "shot" as number of box annotations. I am confused about it.

**Comment by pengyinxw** [comment]
@Tung-I I also find a potential issue when sampling the [negative supports](https://github.com/Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection/blob/eacee935b6fe2a1fd572fa457eb39ae5d6dadb8c/lib/roi_data_layer/fs_loader.py#L143), here num_classes is 81, but only 20 novel categories have data, while the other 61 are empty lists. The condition [here](https://github.com/Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection/blob/eacee935b6fe2a1fd572fa457eb39ae5d6dadb8c/lib/roi_data_layer/fs_loader.py#L144) can not filter out the 61 base categories (+ background), therefore may lead to getting a neg_cls_idx, which doesn't have any data (a empty list).

**Comment by pengyinxw** [comment]
For your provided 10 shots fine-tuning dataset, after[ box filter](https://github.com/Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection/blob/eacee935b6fe2a1fd572fa457eb39ae5d6dadb8c/lib/roi_data_layer/fs_loader.py#L75), some novel categories even have less than 3/5 box annotations in the [support_db](https://github.com/Tung-I/Dual-awareness-Attention-for-Few-shot-Object-Detection/blob/eacee935b6fe2a1fd572fa457eb39ae5d6dadb8c/lib/roi_data_layer/fs_loader.py#L59). E.g. class:05 has 4 box annotations and class:09 only has 2. In this way, even if we set the shot = 3/5, it will still raise the value error: Sample larger than population or is negative.  
Looking forward to your reply. Thank you in advance!

**Comment by Tung-I** [comment]
Sorry for misleading you in the earlier reply. The shot in Tab.1 should be the number of "boxes" rather than "images".  It is embarrassing that I wrongly remembered my experimental setting (number of image v.s. number of boxes).  Therefore, the fine-tuning datasets I prepared and published 2 days ago actually mismatched with my original experimental setting and have been retracted. Here's the code for producing those fine-tuning data (K images per novel class). Hope it could be of some help if you are going generate another kind of ft data (K boxes per novel class): 

https://drive.google.com/file/d/1fo8YTCTrryKYQ-XgT6z8YaUyBNgoT-Gm/view?usp=sharing

**Comment by xiexijun** [comment]
@pengyinxw Thank you very much for your communication with the author that I can learn a lot of knowledge. Could you tell me how to set N and K of the N-way K-shot used in training phase in Table 1. Is it sets to 2-way as in ReadMe or something else. I am not good at English, sorry.

**Comment by Tung-I** [comment]
@xiexijun All of the models in our paper were trained by the 2-way contrastive strategy proposed in [3]. For the "1, 3, 5-way" in our table, it means the number of target classes (in an episode) at inference, which has nothing to do with the training stage. 

@pengyinxw  Thank you for letting me know the fine-tuning datasets I prepared previously would cause bugs. I would find some time to add the well-functioned fine-tuning dataloader and dataset in this repo ASAP. BTW, this is why I proposed to evaluate models under the non-fine-tuning settings in Tab. 2~5: directly testing the models without the fine-tuning stage could be more efficient, less controversial, and more practical in real-world tasks.

I sincerely appreciate your attention to this work.

**Comment by xiexijun** [comment]
@Tung-I Thank you very much for your reply.  I have understood this problem. Based on this, I have another question about how 3 and 5 ways are realized during the inference phase. My personal ability is really too weak, I asked a lot of simple questions, I am sorry to disturb you a lot of time.

**Comment by pengyinxw** [comment]
@Tung-I Thanks for your explanation and provided scripts. It makes sense now.

Another point is for N-way K-shot setting (here K is treated as num of boxes per novel class), I think if we use K box annotations per novel class for fine-tuning, then we should also use theses box annotations as support data to do inference rather than using the cropped images in your provided "supports" folder. This will also violate K-shot setting.
However, it would be fine for the non fine-tuning evaluation protocol.

Again, thanks for your great work and detailed replies!

**Comment by tim20120526** [comment]
> @Tung-I Thanks for your explanation and provided scripts. It makes sense now.
> 
> Another point is for N-way K-shot setting (here K is treated as num of boxes per novel class), I think if we use K box annotations per novel class for fine-tuning, then we should also use theses box annotations as support data to do inference rather than using the cropped images in your provided "supports" folder. This will also violate K-shot setting. However, it would be fine for the non fine-tuning evaluation protocol.
> 
> Again, thanks for your great work and detailed replies!

i also think so
